### ISSUE TITLE
Possible conflict between parameters "fields" and "include"

### ISSUE THREAD

**Issue by hhware** [json-api/json-api#476]
In the section Sparse Fieldsets:

> A client **MAY** request that an endpoint return only specific [fields] in the response on a per-type basis by including a `fields[TYPE]` parameter.

If we use an example from the spec, `GET /articles?include=author&fields[articles]=title,body`, and assume that both `title` and `body` are attributes, does it mean that `author` should not be returned in `links` of articles? 

If yes, it seems to contradict the following:
-  in the section Inclusion of Related Resources: "Each relationship name **MUST** be identical to the key in the links section of its parent resource object",
- in the section Resource Relationships: "If a link object refers to resource objects included in the same compound document, it **MUST** include resource linkage to those resource objects".

Alternatives which I see:
1. Provide a requirement to coordinate values of `include` and `fields` in case the latter includes relationship names, and return 400 if they are not properly coordinated.
2. Change the first quoted phrase to something like: "Client **MAY** request that an endpoint return only specific [attributes] in the response on a per-type basis by including a `fields[TYPE]` parameter."
3. Change the first quoted phrase to something like: "If a compound document is returned, client **MAY** request that an endpoint return only specific [attributes] in the response on a per-type basis by including a `fields[TYPE]` parameter. If a non-compound document is returned, `fields[TYPE]` may additionally include relationship names." Also, add a "Note: In case of compound documents, inclusion of relationship in `links` is effectively managed via parameter `include`."
4. State that inclusion of a relationship name into `include` also automatically adds it to `fields[TYPE]`.

IMHO, the alternative (3) is best because it is most versatile. It will especially make sense for cases when API supports only compound documents -- no ambiguity is possible in this case.

Realistically, servers would probably return either compound or non-compound documents for all endpoints, but I guess regulating it is beyond this scope of this spec.

FYI, a PR on related topic: #471

**Comment by bintoro** [comment]
Good point.

But wouldn't the simplest solution be just to state that `include=author` implies `fields[...]=author`? Clearly that's the intent here.

**Comment by hhware** [comment]
@bintoro, right, the 4th option exists (added it to the initial comment). To me, it is not an attractive option at all: I would prefer the client to do all these string manipulation and provide explicit and noncontradictory instructions to the server. I think such approach is much less error-prone...

PS Sorry, somehow did not add the option (4) to the initial comment, fixed now

**Comment by ethanresnick** [comment]
Hey guys! I thought I commented on here a long time ago, but it looks like Github didn't save my post somehow. Anyway...

> If we use an example from the spec, GET /articles?include=author&fields[articles]=title,body, and assume that both title and body are attributes, does it mean that author should not be returned in links of articles?

My reading of this is also that, yes, this request implies that the authors should show up in the `included` section but not in the links of the articles. Whether this is a reading others want is being discussed in #497. But, if it's a valid reading, then I think there's a fourth option that doesn't limit the scope of `fields` to just filtering attributes or require client coordination of the parameters, namely: simply reword the problematic phrase that @hhware pointed out so that it handles this condition.

So, we'd rewrite:

> Each relationship name MUST be identical to the key in the links section of its parent resource object.

to something like:

> Each relationship name MUST be identical to the key in the links section of its parent resource object, if that key is present. If that key is not present because it has been excluded using sparse fieldsets, the relationship name MUST be identical to the key that would appear were the fieldset restriction removed.

And then the contradiction goes away.

**Comment by hhware** [comment]
Sorry, I misread the clause about the compound documents, it is irrelevant to this. I agree with @ethanresnick that suggested correction would eliminate the existing formal contradiction. It would not eliminate the ambiguity of interpretation, but this is already being discussed in #497.

**Comment by tkellen** [comment]
@hhware, @dgeb and I just discussed this and think that option 1 feels like the most direct way to address this issue. Would you mind opening a PR to make this change?

**Comment by hhware** [comment]
@tkellen, with pleasure!

**Comment by hhware** [comment]
While preparing the PR for it, I realized that option (1) from https://github.com/json-api/json-api/issues/476#issue-61984408 may be interpreted differently. To account for that, and to account for the discussion in #497, I prepared 2 PRs which could resolve this issue:
- #548 requires that if a related resource is included in a compound document, it must be also listed in the links of the resources referring to it (whether those resources are located in primary `data` or in `included`).
- #549 allows the client to specify values of `included` and `fields[TYPE]` independently from each other. The justification for it can be found here: https://github.com/json-api/json-api/issues/497#issuecomment-87162174. If accepted, #549 should also resolve #497.

IMHO, #549 is much better option because it leaves the flexibility to define policies to specific API implementations, while still being strict enough so that libraries could implement the spec in a consistent manner.
